### DISCUSSION TITLE
error about prepared statement in prefork mode

### DISCUSSION THREAD

**Discussion by flash548** [mojolicious/mojo#1976] [Q&A]
Hello, I get an error for `DBD::Pg::st execute failed: ERROR:  prepared statement "dbdpg_p75253_1" already exists.....` (or the like) whenever I run my app using `prefork`.  The error happens within the sub, similar to below:

```
sub get_user ($self, $username) {
  return $self->db->select("users", undef, {email => lc $username})->hash;
}
```

But doesn't occur in just plain old `daemon` mode.  Assume its some handle being left open or something?  The database is Postgres and I'm using Mojo::Pg

**Comment by kraih** [comment]
Never seen that before.

**Reply by flash548** [reply]
yeah I was wondering if it was maybe because the query was left "open" or something - so I tried reading "all" of it (when in reality that query only/should return one result) with `->hashes;` and then index on the returned arrayref.. but that didn't seem to make any difference either ... hmmm

**Reply by flash548** [reply]
Think I might have found the cause of this behavior. It seems this issue was caused by DBD::Pg default behavior to always prepare statements (`$dbh->{pg_server_prepare} = 1`) once on the database server which can/does conflict with that own behavior in other spawned instances (if the library attempts to auto-name a statement that already is stored on the database server).  Workaround to either turn this feature off (`$db_handle->db->dbh->{pg_server_prepare} = 0`) or manually prepare/name the queries ahead of time and then use them amongst the app (amongst the pre-forked instances).
